File: src/riley/pydemos/demo_imagewarp2d.py
# --------------------------------------------------------------------------
# Riley: A High Performance Rasteriser for DIC UQ
#
# Copyright (c) 2025-2026 scepticalrabbit (Lloyd Fletcher)
# Licensed under the MIT License (see LICENSE file for details)
#
# Authors: scepticalrabbit (Lloyd Fletcher)
# --------------------------------------------------------------------------
from __future__ import annotations

from dataclasses import replace
from time import perf_counter

import numpy as np
import riley

from riley.pydemos.common import make_demo_out_dir


def project_nodes_to_uv(
    coords: np.ndarray,
    camera: riley.Camera,
) -> np.ndarray:
    """Project 3D world coordinates to normalized UV space [0, 1] x [0, 1]."""
    alpha_z, beta_y, gamma_x = camera.rot_world
    rot_mat = np.zeros((3, 3), dtype=np.float64)
    rot_mat[0, 0] = np.cos(alpha_z) * np.cos(beta_y)
    rot_mat[0, 1] = (
        np.cos(alpha_z) * np.sin(beta_y) * np.sin(gamma_x)
        - np.sin(alpha_z) * np.cos(gamma_x)
    )
    rot_mat[0, 2] = (
        np.cos(alpha_z) * np.sin(beta_y) * np.cos(gamma_x)
        + np.sin(alpha_z) * np.sin(gamma_x)
    )
    rot_mat[1, 0] = np.sin(alpha_z) * np.cos(beta_y)
    rot_mat[1, 1] = (
        np.sin(alpha_z) * np.sin(beta_y) * np.sin(gamma_x)
        + np.cos(alpha_z) * np.cos(gamma_x)
    )
    rot_mat[1, 2] = (
        np.sin(alpha_z) * np.sin(beta_y) * np.cos(gamma_x)
        - np.cos(alpha_z) * np.sin(gamma_x)
    )
    rot_mat[2, 0] = -np.sin(beta_y)
    rot_mat[2, 1] = np.cos(beta_y) * np.sin(gamma_x)
    rot_mat[2, 2] = np.cos(beta_y) * np.cos(gamma_x)

    pos_w = np.array(camera.pos_world, dtype=np.float64)
    roi_w = np.array(camera.roi_cent_world, dtype=np.float64)
    image_dist = float(np.linalg.norm(pos_w - roi_w))

    sensor_w = camera.pixels_num[0] * camera.pixels_size[0]
    sensor_h = camera.pixels_num[1] * camera.pixels_size[1]
    image_dim_x = (image_dist / camera.focal_length) * sensor_w
    image_dim_y = (image_dist / camera.focal_length) * sensor_h

    c2w = np.eye(4, dtype=np.float64)
    c2w[:3, :3] = rot_mat
    c2w[:3, 3] = pos_w
    w2c = np.linalg.inv(c2w)

    coords_hom = np.hstack([coords, np.ones((coords.shape[0], 1))])
    coords_cam = (w2c @ coords_hom.T).T[:, :3]

    cx = image_dist * coords_cam[:, 0] / (-coords_cam[:, 2])
    cy = image_dist * coords_cam[:, 1] / (-coords_cam[:, 2])

    cx = 2.0 * cx / image_dim_x
    cy = 2.0 * cy / image_dim_y

    u = (cx + 1.0) * 0.5
    v = (1.0 - cy) * 0.5

    uvs = np.zeros((coords.shape[0], 2), dtype=np.float64)
    uvs[:, 0] = u
    uvs[:, 1] = v
    return uvs


def main() -> None:
    data_dir = riley.data.platehole_csv_case_path()
    texture_path = riley.data.speckle_texture_path()
    out_dir = make_demo_out_dir("demo-imagewarp2d")
    pixels_num = (2464, 2056)
    pixels_size = (3.45e-6, 3.45e-6)
    focal_length = 50.0e-3
    fov_scale_factor = 0.65
    sub_sample = 2
    stereo_angle_deg = 20.0
    total_threads = 8
    vertical_translation_y = 0.005  # 5mm vertical shift (half hole diameter)

    distortion_model = {
        "distortion_model": 1,
        "distortion_k1": -0.2,
        "distortion_k2": 0.1,
        "distortion_k3": 0.0,
        "distortion_p1": 0.0001,
        "distortion_p2": -0.0001,
    }

    coords, connect, _, _ = riley.load_sim_csvs(data_dir)

    # 2 Frames: Frame 0 = undeformed reference, Frame 1 = vertical shift
    disp = np.zeros((2, coords.shape[0], 3), dtype=np.float64)
    disp[1, :, 1] = vertical_translation_y

    texture = riley.load_texture_u8(texture_path)

    roi_pos = riley.roi_cent_from_coords(coords)
    camera_0_pos = riley.pos_fill_frame_from_rot(
        coords,
        pixels_num,
        pixels_size,
        focal_length,
        (0.0, 0.0, 0.0),
        fov_scale_factor,
    )
    camera_0 = riley.Camera(
        pixels_num=pixels_num,
        pixels_size=pixels_size,
        pos_world=camera_0_pos,
        rot_world=(0.0, 0.0, 0.0),
        roi_cent_world=roi_pos,
        focal_length=focal_length,
        sub_sample=sub_sample,
        **distortion_model,
    )

    camera_1_rot = (0.0, np.deg2rad(stereo_angle_deg), 0.0)
    camera_1_pos = riley.pos_fill_frame_from_rot(
        coords,
        pixels_num,
        pixels_size,
        focal_length,
        camera_1_rot,
        fov_scale_factor,
    )
    camera_1 = riley.Camera(
        pixels_num=pixels_num,
        pixels_size=pixels_size,
        pos_world=camera_1_pos,
        rot_world=camera_1_rot,
        roi_cent_world=roi_pos,
        focal_length=focal_length,
        sub_sample=sub_sample,
        **distortion_model,
    )

    # Projected UV mapping from reference camera (Camera 0)
    uvs_projected = project_nodes_to_uv(coords, camera_0)

    mesh = riley.Mesh(
        mesh_type=riley.MeshType.quad8,
        coords=coords,
        connect=connect,
        disp=disp,
        shader_type=riley.ShaderType.tex,
        uvs=uvs_projected,
        texture=texture,
        sample=riley.TextureSample.cubic_catmull_rom,
        sample_mode=riley.TextureSampleMode.lut_lerp,
        bits=8,
        scaling_type=riley.ScaleStrategy.none,
    )

    config = riley.create_raster_config(
        num_frames=disp.shape[0],
        total_threads=total_threads,
        save_strategy=riley.SaveStrategy.disk,
    )
    config.background_value = 128.0
    config.tile_size_max = 128
    config.save_scaling = riley.ScaleStrategy.none

    start_time = perf_counter()
    riley.raster([mesh], [camera_0, camera_1], config, out_dir=str(out_dir))
    elapsed_time = perf_counter() - start_time
    print(f"render time: {elapsed_time:.6f} s")

    riley.save_stereo_pair(
        str(out_dir),
        "stereo_data_opengl.csv",
        camera_0,
        camera_1,
    )
    riley.save_stereo_pair(
        str(out_dir),
        "stereo_data_opencv.csv",
        replace(camera_0, coord_sys=riley.CameraCoordSys.opencv),
        replace(camera_1, coord_sys=riley.CameraCoordSys.opencv),
    )
    print(f"rendered imagewarp2d to {out_dir}")


if __name__ == "__main__":
    main()

File: src/demo_imagewarp2d.zig
// --------------------------------------------------------------------------
// Riley: A High Performance Rasteriser for DIC UQ
//
// Copyright (c) 2025-2026 scepticalrabbit (Lloyd Fletcher)
// Licensed under the MIT License (see LICENSE file for details)
//
// Authors: scepticalrabbit (Lloyd Fletcher)
// --------------------------------------------------------------------------
const std = @import("std");

const buildconfig = @import("riley/zig/buildconfig.zig");
const riley = @import("riley/zig/riley.zig");
const RasterConfig = riley.RasterConfig;
const meshio = @import("riley/zig/meshio.zig");
const iio = @import("riley/zig/imageio.zig");
const mo = @import("riley/zig/meshpipeline.zig");
const MeshInput = mo.MeshInput;
const gk = @import("riley/zig/geometrykernels.zig");
const camera_mod = @import("riley/zig/camera.zig");
const cameraio = @import("riley/zig/cameraio.zig");
const cameraops = @import("riley/zig/cameraops.zig");
const sceneops = @import("riley/zig/sceneops.zig");
const Rotation = @import("riley/zig/rotation.zig").Rotation;
const CameraInput = camera_mod.CameraInput;
const DistortionModel = camera_mod.DistortionModel;
const BrownConrady = camera_mod.BrownConrady;
const F = buildconfig.F;

const DATA_DIR = "data/FE/platehole3d_2mr_63f/";
const TEXTURE_PATH = "texture/speckle.bmp";
const OUT_DIR_ROOT = "./out/demo-imagewarp2d";

const PIXELS_NUM = [2]u32{ 2464, 2056 };
const PIXELS_SIZE = [2]F{
    @floatCast(3.45e-6),
    @floatCast(3.45e-6),
};
const FOCAL_LENGTH: F = @floatCast(50.0e-3);
const FOV_SCALE_FACTOR: F = @floatCast(0.65);
const SUB_SAMPLE: u32 = 2;
const STEREO_ANGLE_DEG: F = 20.0;
const VERTICAL_TRANSLATION_Y: F = 0.005; // 5mm vertical shift (half hole diameter)

const TOTAL_THREADS: u16 = 8;
const FRAME_BATCH_SIZE_PER_GROUP: u16 = 1;
const MAX_GEOM_JOBS_IN_FLIGHT_PER_GROUP: u16 = 1;
const RENDER_GROUP_COUNT: usize = 8;
const WORKERS_PER_GROUP: u16 = 1;

fn buildDistortion() DistortionModel {
    return .{ .brown_conrady = BrownConrady{
        .k1 = -0.2,
        .k2 = 0.1,
        .k3 = 0.0,
        .p1 = 0.0001,
        .p2 = -0.0001,
    } };
}

pub fn main(init: std.process.Init) !void {
    const outer_alloc = init.gpa;

    var arena = std.heap.ArenaAllocator.init(outer_alloc);
    defer arena.deinit();
    const aa = arena.allocator();

    // 1. Setup Rasteriser Configuration
    const config = RasterConfig{
        .render_mode = .offline,
        .total_threads = TOTAL_THREADS,
        .frame_batch_size_per_group = FRAME_BATCH_SIZE_PER_GROUP,
        .max_geom_jobs_in_flight_per_group = MAX_GEOM_JOBS_IN_FLIGHT_PER_GROUP,
        .max_geom_workers_per_job = 1,
        .geom_scheduling_mode = .spread,
        .max_raster_workers_per_job = 1,
        .save_strategy = .disk,
        .tile_size_min = 8,
        .tile_size_max = 128,
        .background_value = 128.0,
        .image_save_opts = &[_]iio.ImageSaveOpts{
            .{ .format = .bmp, .bits = 8, .scaling = .none },
        },
        .report = .bench,
    };

    const managed_ios = try outer_alloc.alloc(std.Io.Threaded, RENDER_GROUP_COUNT);
    defer {
        for (managed_ios) |*managed_io| {
            managed_io.deinit();
        }
        outer_alloc.free(managed_ios);
    }

    const render_groups = try outer_alloc.alloc(
        riley.RenderGroupSpec,
        RENDER_GROUP_COUNT,
    );
    defer outer_alloc.free(render_groups);

    for (0..RENDER_GROUP_COUNT) |gg| {
        managed_ios[gg] = riley.getThreadedIo(
            aa,
            init.minimal,
            WORKERS_PER_GROUP,
        );
        render_groups[gg] = .{
            .io = managed_ios[gg].io(),
            .workers = WORKERS_PER_GROUP,
        };
    }
    const io = render_groups[0].io;

    // 2. Load Mesh Simulation Geometry
    std.debug.print(
        "Loading plate with hole mesh from {s}...\n",
        .{DATA_DIR},
    );
    const coord_path = DATA_DIR ++ "coords.csv";
    const conn_path = DATA_DIR ++ "connect.csv";

    const sim_data = try meshio.loadSimData(
        aa,
        io,
        coord_path,
        conn_path,
        null,
        null,
    );

    // 3. Create Custom Displacement Field (2 frames: 0 disp, then vertical shift)
    const node_count = sim_data.coords.mat.rows_num;
    var disp_field = try meshio.Field.initAlloc(aa, 2, node_count, 3);
    for (0..node_count) |nn| {
        // Frame 0: undeformed reference
        disp_field.array.set(&[_]usize{ 0, nn, 0 }, 0.0);
        disp_field.array.set(&[_]usize{ 0, nn, 1 }, 0.0);
        disp_field.array.set(&[_]usize{ 0, nn, 2 }, 0.0);

        // Frame 1: translated vertically along the long axis (Y)
        disp_field.array.set(&[_]usize{ 1, nn, 0 }, 0.0);
        disp_field.array.set(&[_]usize{ 1, nn, 1 }, VERTICAL_TRANSLATION_Y);
        disp_field.array.set(&[_]usize{ 1, nn, 2 }, 0.0);
    }

    // 4. Load Texture for Projected Warping
    std.debug.print("Loading speckle texture from {s}...\n", .{TEXTURE_PATH});
    const texture = try iio.loadImage(
        u8,
        1,
        aa,
        io,
        TEXTURE_PATH,
        .bmp,
    );

    // 5. Setup Cameras (Stereo Rig)
    std.debug.print("Setting up stereo camera pair...\n", .{});
    const roi_pos = sceneops.boundsCenter(&sim_data.coords);
    const distortion = buildDistortion();

    // Camera 0: face on (reference camera)
    const cam0_rot = Rotation.init(
        std.math.degreesToRadians(0.0),
        std.math.degreesToRadians(0.0),
        std.math.degreesToRadians(0.0),
    );
    const cam0_pos = cameraops.posFillFrameFromRot(
        &sim_data.coords,
        PIXELS_NUM,
        PIXELS_SIZE,
        FOCAL_LENGTH,
        cam0_rot,
        FOV_SCALE_FACTOR,
    );
    const cam0_in = CameraInput{
        .pixels_num = PIXELS_NUM,
        .pixels_size = PIXELS_SIZE,
        .pos_world = cam0_pos,
        .rot_world = cam0_rot,
        .roi_cent_world = roi_pos,
        .focal_length = FOCAL_LENGTH,
        .sub_sample = SUB_SAMPLE,
        .distortion = distortion,
    };

    // Camera 1: stereo angle
    const cam1_rot = Rotation.init(
        std.math.degreesToRadians(0.0),
        std.math.degreesToRadians(STEREO_ANGLE_DEG),
        std.math.degreesToRadians(0.0),
    );
    const cam1_pos = cameraops.posFillFrameFromRot(
        &sim_data.coords,
        PIXELS_NUM,
        PIXELS_SIZE,
        FOCAL_LENGTH,
        cam1_rot,
        FOV_SCALE_FACTOR,
    );
    const cam1_in = CameraInput{
        .pixels_num = PIXELS_NUM,
        .pixels_size = PIXELS_SIZE,
        .pos_world = cam1_pos,
        .rot_world = cam1_rot,
        .roi_cent_world = roi_pos,
        .focal_length = FOCAL_LENGTH,
        .sub_sample = SUB_SAMPLE,
        .distortion = distortion,
    };

    // 6. Prepare Mesh Input with ProjectedTex Shader
    std.debug.print("Configuring ProjectedTex 2D image warper...\n", .{});
    const mesh_input = MeshInput{
        .mesh_type = .quad8,
        .coords = sim_data.coords,
        .connect = sim_data.connect,
        .disp = disp_field,
        .shader = .{
            .projected_tex_u8 = .{
                .ref_camera = cam0_in,
                .ref_coords = null,
                .tex = texture,
            },
        },
    };

    // 7. Run the Rasteriser
    std.debug.print(
        "Rendering projected 2D image warping to {s}/...\n",
        .{OUT_DIR_ROOT},
    );
    const meshes = [_]MeshInput{mesh_input};
    const cams_in = [_]CameraInput{ cam0_in, cam1_in };

    std.Io.Dir.cwd().deleteTree(io, OUT_DIR_ROOT) catch |err| {
        if (err != error.FileNotFound) return err;
    };

    const images = try riley.raster(
        aa,
        render_groups,
        &cams_in,
        &meshes,
        config,
        OUT_DIR_ROOT,
    );

    if (images) |img| {
        aa.free(img.slice);
        img.deinit(aa);
    }

    var out_dir = try std.Io.Dir.cwd().openDir(io, OUT_DIR_ROOT, .{});
    defer out_dir.close(io);

    var cam0_opengl = cam0_in;
    cam0_opengl.coord_sys = .opengl;
    var cam1_opengl = cam1_in;
    cam1_opengl.coord_sys = .opengl;
    try cameraio.saveStereoPair(
        io,
        out_dir,
        "stereo_data_opengl.csv",
        .{ .cameras = .{ cam0_opengl, cam1_opengl } },
    );

    var cam0_opencv = cam0_in;
    cam0_opencv.coord_sys = .opencv;
    var cam1_opencv = cam1_in;
    cam1_opencv.coord_sys = .opencv;
    try cameraio.saveStereoPair(
        io,
        out_dir,
        "stereo_data_opencv.csv",
        .{ .cameras = .{ cam0_opencv, cam1_opencv } },
    );

    std.debug.print(
        "Demo complete. Warped images saved to {s}/\n",
        .{OUT_DIR_ROOT},
    );
}
